lib/client.js

```diff
@@ -583,9 +583,13 @@ MqttClient.prototype.end = function (force, cb) {
 
   function closeStores () {
     that.disconnected = true
-    that.incomingStore.close(function () {
-      that.outgoingStore.close(cb)
-    })
+    if (that.options.clean) {
```

Comment:
this is not correct. Some of the stores close connections here. If we do not call `close()`, we will not release those resources and quit cleanly.

Reply:
I think that MQTT Clean Session flag is corresponding to `options.clean`.
MQTT Clean Session is defined as follows:
See http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html#_Toc385349231
> If CleanSession is set to 0, the Server MUST resume communications with the Client based on state from the current Session (as identified by the Client identifier). If there is no Session associated with the Client identifier the Server MUST create a new Session. The Client and Server MUST store the Session after the Client and Server are disconnected [MQTT-3.1.2-4]. After the disconnection of a Session that had CleanSession set to 0, the Server MUST store further QoS 1 and QoS 2 messages that match any subscriptions that the client had at the time of disconnection as part of the Session state [MQTT-3.1.2-5]. It MAY also store QoS 0 messages that meet the same criteria.

That means the client need to keep the stores after disconnect sending if CleanSssion flag is false.
What do you think?

Reply:
It needs to _keep_ the content of the stores, but the default are in-memory and we are clearing the context during close: https://github.com/mqttjs/MQTT.js/blob/master/lib/store.js#L103. That might be wrong. Close should really be a no-op by default on those.

If you consider only MQTT, you are correct. However, we need to take into account that this is a node module. If the stores are maintaining a connection to a database, those normally prevent the process to terminate correctly. So, `close()` needs to __always be called to free those resources__.

Reply:
I understand the situation. I started thinking how to achieve both. 
As you mentioned, add a flag seems good.
IIUC, adding the clean flag to `mqtt.Client#end()`. So the interface is updated to `end([force=false], [clean=false], [cb])`.
If it is ok, I will add the commit.

Reply:
I wrote `[clean=false]` but it should be `[clean=true]` for the default behavior.

Reply:
If `that.options.clean` is true, stores should be cleaned even if `end()`'s clean is false.
So the code will be as follows:
```cpp
MqttClient.prototype.end = function (force, clean, cb) {
  var that = this

  if (typeof force === 'function') {
    cb = force
    force = false
    clean = true
  } else if (typeof clean === 'function') {
    cb = clean
    clean = true
  }

  function closeStores () {
    if (clean || that.options.clean) {
      that.disconnected = true
      that.incomingStore.close(function () {
        that.outgoingStore.close(cb)
      })
    } else {
      if (cb) cb()
    }
  }
  ...
```

Reply:
I'm not sure the default parameter policy. But it seems that if users omit the parameter, the parameter regard to false in most cases. So I think that it is better to replace `[clean=true]` with `[keep=false]`. What do you think?

Reply:
Unfortunately, I’m not able to convey my reasoning: the close() method is something that MUST happen in the normal case independently of the MQTT clean flag. We should just add another option closeStoreOnEnd: false, so that in your edge case you can avoid calling close. This means that the stores would need to be closed by the application and they are not managed anymore by MQTT.js.

Reply:
I'm trying to understand what you mean. Now, I got some idea. 
Here is what I understand:
* `Store#close()` must always be called in `MqttClient#end()` because it is a kind of **contract**.
* Users might prepare a custom Store, and they need to chance to cleanup resources in user custom Store.

So I consider some options.

My first idea is passing `that.options.clean` to `Store#close()`.

```cpp
MqttClient.prototype.end = function (force, cb) {
  var that = this

  if (typeof force === 'function') {
    cb = force
    force = false
  }

  function closeStores () {
    that.disconnected = true
    that.incomingStore.close(that.options.clean, function () {
      that.outgoingStore.close(that.options.clean, cb)
    })
  }
```

`Store` can change behavior using the flag `that.options.clean`. I think it is good. *1

However, it breaks existing code because it changes `Store#close()` interface. I think it is acceptable. But it is NOT acceptable, we can change the order of the parameter like 

```cpp
that.outgoingStore.close(cb, that.options.clean)
```

It seems to be inconsistent. All other functions' callback is located on the last.

Another option is adding options to `Store`'s constructor. We can pass the same value as `options.clean` of `MqttClient`. I think that it is a little more confusing than *1.

What do you think?

Reply:
'Store' (on memory)  is modified as follows:

```cpp
/**
 * Close the store
 */
Store.prototype.close = function (clean, cb) {
  if (clean) {
    this._inflights = null
  }
  if (cb) {
    cb()
  }
```

Reply:
I'm 👍  in passing `options.clean` to the store. However, pass it on creation rather than during `close()`, so we don't have to change the API.

Reply:
Thank you! I added a commit. 
I guess you think that MQTT's clean session and `Store()` clean are different concept.
So I don't pass MQTT's clean session to `Store()`'s constructor at
https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L88

I can control `Store`'s  behavior as https://github.com/mqttjs/MQTT.js/pull/707/files#diff-16702ac352d8772fe21f4f9c1de7e900R1961

Reply:
> So I don't pass MQTT's clean session to Store()'s constructor at
> https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L88

If you think MQTT's clean session should be passed to Store()'s constructor, I will update the code.